aten/src/ATen/native/ConvUtils.h

```diff
@@ -214,4 +218,179 @@ static inline bool miopen_conv_use_channels_last(const at::Tensor& input, const
   return can_use_miopen_channels_last_2d || can_use_miopen_channels_last_3d;
 }
 
+static Tensor view_weight_2d(const Tensor& weight_) {
+  Tensor weight = weight_.contiguous();
+  if (weight.dim() == 4) {
+    const int64_t s1 = weight.size(0);
+    const int64_t s2 = weight.size(1) * weight.size(2) * weight.size(3);
+    return weight.view({s1, s2});
+  } else {
+    return weight;
+  }
+}
+
+// Computes finput used in slow 2D kernel computation.
+// This is computed separately in the forward and backward passes.
+static Tensor compute_finput2d(
```

Comment:
> finput is now recomputed in the backward pass for the slow 2d / 3d kernels instead of being saved from the forward pass. The logic for is based on the forward computation and is present in compute_finput2d / compute_finput3d functions in ConvUtils.h.

This lead to a performance slowdown, right? Do we know how bad it is and if it is OK to have?

Reply:
I can get you some details on the magnitude of the slowdown. I mentioned this to Ed before and his stance was more or less "it's already the slow algorithm; don't worry about it"

Reply:
Here's some rough benchmark results, measured in seconds, averaged over 100 forward / backward runs each (10 each for the last two cases):

|  | Old convolution | with convolution_backward |
| --- | --- | --- |
| Input shape (2, 3, 10) | 0.0001408773532602936s (std: 8.87400601641275e-05) | 0.00013880817277822644s (std: 9.332414629170671e-05) |
| Input shape (2, 3, 100) | 0.00013632909394800663s (std: 2.9321088732103817e-05) | 0.0001252235088031739s (std: 1.944660471053794e-05) |
| Input shape (2, 3, 1000) | 0.00016042383504100144s (std: 2.2567743144463748e-05) | 0.00017399227363057435s (std: 4.868006726610474e-05) |
| Input shape (2, 3, 10, 10) | 0.00012011521175736561s (std: 3.10868417727761e-05) | 0.0001369411329505965s (std: 4.33642890129704e-05) |
| Input shape (2, 3, 100, 100) | 0.001363431103527546s (std: 0.0002012234617723152) | 0.002269844990223646s (std: 0.0004060452920384705) |
| Input shape (2, 3, 1000, 1000) | 0.15414561331272125s (std: 0.00891160313040018) | 0.23099897801876068s (std: 0.010208758525550365) |
| Input shape (2, 3, 10, 10, 10) | 0.0005283193895593286s (std: 0.00036496721440926194) | 0.0006224568351171911s (std: 0.0003632625157479197) |
| Input shape (2, 3, 100, 100, 100) | 0.571362316608429s (std: 0.014484521932899952) | 0.6644342541694641s (std: 0.012034161016345024) |
| Input shape (2, 3, 3, 578, 578) | 0.19007597863674164s (std: 0.007472589146345854) | 0.21315564215183258s (std: 0.007333348970860243) |

Reply:
Results on CUDA averaged over 100 forward / backward runs each (measured in sec):
|  | Old convolution | with convolution_backward |
| --- | --- | --- |
| Input shape (2, 3, 10, 10) | 0.00025939932675100863s (std: 1.153016637545079e-05) | 0.0003229645953979343s (std: 1.5169899597822223e-05) |
| Input shape (2, 3, 100, 100) | 0.00027780578238889575s (std: 1.019740739138797e-05) | 0.00034236189094372094s (std: 1.1811538570327684e-05) |
| Input shape (2, 3, 1000, 1000) | 0.003029137384146452s (std: 5.504746695805807e-06) | 0.003656409913673997s (std: 5.408414381236071e-06) |
| Input shape (2, 3, 10000, 10000) | 0.27684372663497925s (std: 0.00010143844701815397) | 0.33951789140701294s (std: 5.043909186497331e-05) |

Reply:
I tried something interesting - instead of using the Slow2d backend during the backward pass for CUDA, use the SlowDilated2d backend. This actually gives less of a performance regression than manually recomputing finput before calling the op. New results:

|  | Old convolution | with convolution_backward |
| --- | --- | --- |
| Input shape (2, 3, 10, 10) | 0.00025939932675100863s (std: 1.153016637545079e-05) | 0.00028246117290109396s (std: 1.8273476598551497e-05) |
| Input shape (2, 3, 100, 100) | 0.00027780578238889575s (std: 1.019740739138797e-05) | 0.00031026892247609794s (std: 1.442296434106538e-05) |
| Input shape (2, 3, 1000, 1000) | 0.003029137384146452s (std: 5.504746695805807e-06) | 0.0030730068683624268s (std: 0.00034801315632648766) |
| Input shape (2, 3, 10000, 10000) | 0.27684372663497925s (std: 0.00010143844701815397) | 0.27632075548171997s (std: 5.34579849045258e-05) |

Reply:
What's the difference between slow2d and slowDilated2d?

Reply:
That is a great question! After some exploration, it looks like slow2d on CUDA recomputes the column data and ignores what is passed in entirely. (OTOH the CPU implementation does use the passed-in column data).

So TL;DR: recomputing column data manually for CUDA is pointless overhead that just wastes time. SlowDilated2d doesn't do this, hence why it's faster.

Reply:
Update: I changed the logic to pass an empty, properly-shaped tensor to slow2d and the perf regression is gone:

|  | Old convolution | with convolution_backward |
| --- | --- | --- |
| Input shape (2, 3, 10, 10) | 0.00025939932675100863s (std: 1.153016637545079e-05) | 0.0002598063147161156s (std: 1.0754698450909927e-05) |
| Input shape (2, 3, 100, 100) | 0.00027780578238889575s (std: 1.019740739138797e-05) | 0.00028062431374564767s (std: 1.1466747309896164e-05) |
| Input shape (2, 3, 1000, 1000) | 0.003029137384146452s (std: 5.504746695805807e-06) | 0.003027678234502673s (std: 6.1277664826775435e-06) |
| Input shape (2, 3, 10000, 10000) | 0.27684372663497925s (std: 0.00010143844701815397) | 0.2763238251209259s (std: 0.0001723081077216193) |